### ISSUE TITLE
Inconsistent `run_spec` during zarr dataset opening by multiple bound methods

### ISSUE THREAD

**Issue by templiert** [pydata/xarray#9325]
### What happened?

A dask distributed client lets multiple agents read data from a zarr store:
```python
for agent in agents:
    client.submit(agent.read_data)
```
If the agents load the same data, then there is a run_spec inconsistency.

I am experiencing deadlocks with many agents and tasks in an application running 24/7, and as mentioned by the following warning message referring to [this issue](https://github.com/dask/dask/issues/9888), I suspect that this `run_spec` inconsistency is the cause of the deadlocks:

```
WARNING | distributed.scheduler | Detected different `run_spec` for key
'original-open_dataset-b-5a5934048f8680ef6d38726770b86391' 
between two consecutive calls to `update_graph`. 
This can cause failures and deadlocks down the line. Please ensure unique key names. 
If you are using a standard dask collections, consider releasing all the data before resubmitting another computation. 
More details and help can be found at https://github.com/dask/dask/issues/9888.

Debugging information
---------------------
old task state: memory
old run_spec: (<function execute_task at 0x000002C3869591C0>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x000002C3886C7900>, key=BasicIndexer((slice(None, None, None),))))),), {})
new run_spec: (<function execute_task at 0x000002C3869591C0>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x000002C3886D9F00>, key=BasicIndexer((slice(None, None, None),))))),), {})
old token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('2a84eba15180db3d422384d14b52d55c8ec9c4fd', ['05fe405753166f125559e7c9ac558654f107c7e9'])]), ('dict', [])])
new token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('2a84eba15180db3d422384d14b52d55c8ec9c4fd', ['2da998e53403d195a27e3e251b85adb7ecb9e29d'])]), ('dict', [])])
old dependencies: set()
new dependencies: set()
```

I believe it is a reasonable workflow to let agents read data independently from disk. Because we discover over time what data is needed, I cannot avoid that multiple agents may redundantly read the same data.

If `read_data` is the bound method of an agent, then the `run_spec` issue occurs.
If `read_data` is not a bound method, then no issue. The log shows a proper release of the keys in that case:
```
DEBUG | distributed.scheduler | Client Client-worker-ac3b4576-5519-11ef-8a98-089204dc4273 releases keys: [('open_dataset-y-c34960ff895a9a4a2d1c966c401b7d40', 0)]
```

### What did you expect to happen?

No `run_spec` issue.

### Minimal Complete Verifiable Example

```Python
from __future__ import annotations

import logging
import sys
from dataclasses import dataclass
from pathlib import Path

import xarray as xr
from distributed import Client


@dataclass
class Agent:
    """Agent."""

    def _read_data(self, path: Path) -> None:
        """Reads data from xarray saved as .zarr."""
        read_data(path=path)


def read_data(path: Path) -> None:
    """Reads data from xarray saved as .zarr."""
    xr.open_zarr(path).load()


def save_dataset_zarr(path: Path) -> None:
    """Saves a toy dataset to disk in .zarr format."""
    ds = xr.Dataset(
        data_vars={"a": ("x", [5, 7]), "b": ("x", [0.1, 2.4])},
        coords={"x": ["a", "b"], "y": ("x", [0, 1])},
    )
    ds.to_zarr(path, mode="a")
    

def set_distributed_loggers_to_debug() -> None:
    """Sets logging level of distributed loggers to debug."""
    handler = logging.StreamHandler(sys.stdout)
    handler.setFormatter(
        logging.Formatter("{levelname:^7} | {name:<35} | {message}", style="{")
    )
    for name_logger, logger in logging.Logger.manager.loggerDict.items():
        if hasattr(logger, "setLevel") and (
            "distributed" in name_logger
            or (hasattr(logger, "name") and "distributed" in logger.name)
        ):
            logger.setLevel(logging.DEBUG)
            logger.addHandler(handler)


if __name__ == "__main__":
    path_zarr = Path(r"C:\tests\dataset.zarr")
    save_dataset_zarr(path=path_zarr)
    REPEATS = range(2)

    with Client(processes=True) as client:
        client.run_on_scheduler(set_distributed_loggers_to_debug)
        client.run(set_distributed_loggers_to_debug, nanny=True)

        print("***\nsubmit read data tasks of agents\n***")
        futures = [
            client.submit(Agent()._read_data, path=path_zarr) for repeat in REPEATS
        ]
        client.gather(futures)  # <- run_spec issue with 2 repeats

        print("***\nsubmit read data tasks\n***")
        futures = [client.submit(read_data, path=path_zarr) for repeat in REPEATS]
        client.gather(futures)  # <- no run_spec issue with many repeats
```

### MVCE confirmation

- [x] Minimal example — the example is as focused as reasonably possible to demonstrate the underlying issue in xarray.
- [x] Complete example — the example is self-contained, including all data and the text of any traceback.
- [x] Verifiable example — the example copy & pastes into an IPython prompt or [Binder notebook](https://mybinder.org/v2/gh/pydata/xarray/main?urlpath=lab/tree/doc/examples/blank_template.ipynb), returning the result.
- [x] New issue — a search of GitHub Issues suggests this is not a duplicate.
- [x] Recent environment — the issue occurs with the latest version of xarray and its dependencies.

### Relevant log output

```Python
WARNING | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-a-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`. This can cause failures and deadlocks down the line. Please ensure unique key names. If you are using a standard dask collections, consider releasing all the data before resubmitting another computation. More details and help can be found at https://github.com/dask/dask/issues/9888.
Debugging information
---------------------
old task state: processing
old run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDD7A6C0>, key=BasicIndexer((slice(None, None, None),))))),), {})
new run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDFF9540>, key=BasicIndexer((slice(None, None, None),))))),), {})
old token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['8473dc8afac710c0d78f4dadf0eb1b8ef9fe5c46'])]), ('dict', [])])
new token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['05fe405753166f125559e7c9ac558654f107c7e9'])]), ('dict', [])])
old dependencies: set()
new dependencies: set()

WARNING | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-a-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`. This can cause failures and deadlocks down the line. Please ensure unique key names. If you are using a standard dask collections, consider releasing all the data before resubmitting another computation. More details and help can be found at https://github.com/dask/dask/issues/9888.
Debugging information
---------------------
old task state: processing
old run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDD7A6C0>, key=BasicIndexer((slice(None, None, None),))))),), {})
new run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDFF9540>, key=BasicIndexer((slice(None, None, None),))))),), {})
old token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['8473dc8afac710c0d78f4dadf0eb1b8ef9fe5c46'])]), ('dict', [])])
new token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['05fe405753166f125559e7c9ac558654f107c7e9'])]), ('dict', [])])
old dependencies: set()
new dependencies: set()

WARNING | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-a-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`. This can cause failures and deadlocks down the line. Please ensure unique key names. If you are using a standard dask collections, consider releasing all the data before resubmitting another computation. More details and help can be found at https://github.com/dask/dask/issues/9888.
Debugging information
---------------------
old task state: processing
old run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDD7A6C0>, key=BasicIndexer((slice(None, None, None),))))),), {})
new run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDFF9540>, key=BasicIndexer((slice(None, None, None),))))),), {})
old token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['8473dc8afac710c0d78f4dadf0eb1b8ef9fe5c46'])]), ('dict', [])])
new token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['05fe405753166f125559e7c9ac558654f107c7e9'])]), ('dict', [])])
old dependencies: set()
new dependencies: set()

WARNING | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-a-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`. This can cause failures and deadlocks down the line. Please ensure unique key names. If you are using a standard dask collections, consider releasing all the data before resubmitting another computation. More details and help can be found at https://github.com/dask/dask/issues/9888.
Debugging information
---------------------
old task state: processing
old run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDD7A6C0>, key=BasicIndexer((slice(None, None, None),))))),), {})
new run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDFF9540>, key=BasicIndexer((slice(None, None, None),))))),), {})
old token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['8473dc8afac710c0d78f4dadf0eb1b8ef9fe5c46'])]), ('dict', [])])
new token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['05fe405753166f125559e7c9ac558654f107c7e9'])]), ('dict', [])])
old dependencies: set()
new dependencies: set()

WARNING | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-a-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`. This can cause failures and deadlocks down the line. Please ensure unique key names. If you are using a standard dask collections, consider releasing all the data before resubmitting another computation. More details and help can be found at https://github.com/dask/dask/issues/9888.
Debugging information
---------------------
old task state: processing
old run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDD7A6C0>, key=BasicIndexer((slice(None, None, None),))))),), {})
new run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDFF9540>, key=BasicIndexer((slice(None, None, None),))))),), {})
old token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['8473dc8afac710c0d78f4dadf0eb1b8ef9fe5c46'])]), ('dict', [])])
new token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['05fe405753166f125559e7c9ac558654f107c7e9'])]), ('dict', [])])
old dependencies: set()
new dependencies: set()

WARNING | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-a-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`. This can cause failures and deadlocks down the line. Please ensure unique key names. If you are using a standard dask collections, consider releasing all the data before resubmitting another computation. More details and help can be found at https://github.com/dask/dask/issues/9888.
Debugging information
---------------------
old task state: processing
old run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDD7A6C0>, key=BasicIndexer((slice(None, None, None),))))),), {})
new run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDFF9540>, key=BasicIndexer((slice(None, None, None),))))),), {})
old token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['8473dc8afac710c0d78f4dadf0eb1b8ef9fe5c46'])]), ('dict', [])])
new token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['05fe405753166f125559e7c9ac558654f107c7e9'])]), ('dict', [])])
old dependencies: set()
new dependencies: set()

WARNING | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-a-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`. This can cause failures and deadlocks down the line. Please ensure unique key names. If you are using a standard dask collections, consider releasing all the data before resubmitting another computation. More details and help can be found at https://github.com/dask/dask/issues/9888.
Debugging information
---------------------
old task state: processing
old run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDD7A6C0>, key=BasicIndexer((slice(None, None, None),))))),), {})
new run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDFF9540>, key=BasicIndexer((slice(None, None, None),))))),), {})
old token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['8473dc8afac710c0d78f4dadf0eb1b8ef9fe5c46'])]), ('dict', [])])
new token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['05fe405753166f125559e7c9ac558654f107c7e9'])]), ('dict', [])])
old dependencies: set()
new dependencies: set()

WARNING | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-a-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`. This can cause failures and deadlocks down the line. Please ensure unique key names. If you are using a standard dask collections, consider releasing all the data before resubmitting another computation. More details and help can be found at https://github.com/dask/dask/issues/9888.
Debugging information
---------------------
old task state: processing
old run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDD7A6C0>, key=BasicIndexer((slice(None, None, None),))))),), {})
new run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDFF9540>, key=BasicIndexer((slice(None, None, None),))))),), {})
old token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['8473dc8afac710c0d78f4dadf0eb1b8ef9fe5c46'])]), ('dict', [])])
new token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['05fe405753166f125559e7c9ac558654f107c7e9'])]), ('dict', [])])
old dependencies: set()
new dependencies: set()

WARNING | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-a-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`. This can cause failures and deadlocks down the line. Please ensure unique key names. If you are using a standard dask collections, consider releasing all the data before resubmitting another computation. More details and help can be found at https://github.com/dask/dask/issues/9888.
Debugging information
---------------------
old task state: processing
old run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDD7A6C0>, key=BasicIndexer((slice(None, None, None),))))),), {})
new run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDFF9540>, key=BasicIndexer((slice(None, None, None),))))),), {})
old token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['8473dc8afac710c0d78f4dadf0eb1b8ef9fe5c46'])]), ('dict', [])])
new token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['05fe405753166f125559e7c9ac558654f107c7e9'])]), ('dict', [])])
old dependencies: set()
new dependencies: set()

2024-08-07 20:03:42,747 - distributed.scheduler - WARNING - Detected different `run_spec` for key 'original-open_dataset-a-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`. This can cause failures and deadlocks down the line. Please ensure unique key names. If you are using a standard dask collections, consider releasing all the data before resubmitting another computation. More details and help can be found at https://github.com/dask/dask/issues/9888.
Debugging information
---------------------
old task state: processing
old run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDD7A6C0>, key=BasicIndexer((slice(None, None, None),))))),), {})
new run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDFF9540>, key=BasicIndexer((slice(None, None, None),))))),), {})
old token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['8473dc8afac710c0d78f4dadf0eb1b8ef9fe5c46'])]), ('dict', [])])
new token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['05fe405753166f125559e7c9ac558654f107c7e9'])]), ('dict', [])])
old dependencies: set()
new dependencies: set()

WARNING | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-a-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`. This can cause failures and deadlocks down the line. Please ensure unique key names. If you are using a standard dask collections, consider releasing all the data before resubmitting another computation. More details and help can be found at https://github.com/dask/dask/issues/9888.
Debugging information
---------------------
old task state: processing
old run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDD7A6C0>, key=BasicIndexer((slice(None, None, None),))))),), {})
new run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDFF9540>, key=BasicIndexer((slice(None, None, None),))))),), {})
old token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['8473dc8afac710c0d78f4dadf0eb1b8ef9fe5c46'])]), ('dict', [])])
new token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['05fe405753166f125559e7c9ac558654f107c7e9'])]), ('dict', [])])
old dependencies: set()
new dependencies: set()

WARNING | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-a-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`. This can cause failures and deadlocks down the line. Please ensure unique key names. If you are using a standard dask collections, consider releasing all the data before resubmitting another computation. More details and help can be found at https://github.com/dask/dask/issues/9888.
Debugging information
---------------------
old task state: processing
old run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDD7A6C0>, key=BasicIndexer((slice(None, None, None),))))),), {})
new run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDFF9540>, key=BasicIndexer((slice(None, None, None),))))),), {})
old token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['8473dc8afac710c0d78f4dadf0eb1b8ef9fe5c46'])]), ('dict', [])])
new token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['05fe405753166f125559e7c9ac558654f107c7e9'])]), ('dict', [])])
old dependencies: set()
new dependencies: set()

WARNING | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-a-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`. This can cause failures and deadlocks down the line. Please ensure unique key names. If you are using a standard dask collections, consider releasing all the data before resubmitting another computation. More details and help can be found at https://github.com/dask/dask/issues/9888.
Debugging information
---------------------
old task state: processing
old run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDD7A6C0>, key=BasicIndexer((slice(None, None, None),))))),), {})
new run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDFF9540>, key=BasicIndexer((slice(None, None, None),))))),), {})
old token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['8473dc8afac710c0d78f4dadf0eb1b8ef9fe5c46'])]), ('dict', [])])
new token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['05fe405753166f125559e7c9ac558654f107c7e9'])]), ('dict', [])])
old dependencies: set()
new dependencies: set()

WARNING | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-a-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`. This can cause failures and deadlocks down the line. Please ensure unique key names. If you are using a standard dask collections, consider releasing all the data before resubmitting another computation. More details and help can be found at https://github.com/dask/dask/issues/9888.
Debugging information
---------------------
old task state: processing
old run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDD7A6C0>, key=BasicIndexer((slice(None, None, None),))))),), {})
new run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDFF9540>, key=BasicIndexer((slice(None, None, None),))))),), {})
old token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['8473dc8afac710c0d78f4dadf0eb1b8ef9fe5c46'])]), ('dict', [])])
new token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['05fe405753166f125559e7c9ac558654f107c7e9'])]), ('dict', [])])
old dependencies: set()
new dependencies: set()

WARNING | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-a-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`. This can cause failures and deadlocks down the line. Please ensure unique key names. If you are using a standard dask collections, consider releasing all the data before resubmitting another computation. More details and help can be found at https://github.com/dask/dask/issues/9888.
Debugging information
---------------------
old task state: processing
old run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDD7A6C0>, key=BasicIndexer((slice(None, None, None),))))),), {})
new run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDFF9540>, key=BasicIndexer((slice(None, None, None),))))),), {})
old token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['8473dc8afac710c0d78f4dadf0eb1b8ef9fe5c46'])]), ('dict', [])])
new token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['05fe405753166f125559e7c9ac558654f107c7e9'])]), ('dict', [])])
old dependencies: set()
new dependencies: set()

WARNING | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-a-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`. This can cause failures and deadlocks down the line. Please ensure unique key names. If you are using a standard dask collections, consider releasing all the data before resubmitting another computation. More details and help can be found at https://github.com/dask/dask/issues/9888.
Debugging information
---------------------
old task state: processing
old run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDD7A6C0>, key=BasicIndexer((slice(None, None, None),))))),), {})
new run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDFF9540>, key=BasicIndexer((slice(None, None, None),))))),), {})
old token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['8473dc8afac710c0d78f4dadf0eb1b8ef9fe5c46'])]), ('dict', [])])
new token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['05fe405753166f125559e7c9ac558654f107c7e9'])]), ('dict', [])])
old dependencies: set()
new dependencies: set()

WARNING | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-a-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`. This can cause failures and deadlocks down the line. Please ensure unique key names. If you are using a standard dask collections, consider releasing all the data before resubmitting another computation. More details and help can be found at https://github.com/dask/dask/issues/9888.
Debugging information
---------------------
old task state: processing
old run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDD7A6C0>, key=BasicIndexer((slice(None, None, None),))))),), {})
new run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDFF9540>, key=BasicIndexer((slice(None, None, None),))))),), {})
old token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['8473dc8afac710c0d78f4dadf0eb1b8ef9fe5c46'])]), ('dict', [])])
new token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['05fe405753166f125559e7c9ac558654f107c7e9'])]), ('dict', [])])
old dependencies: set()
new dependencies: set()

WARNING | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-a-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`. This can cause failures and deadlocks down the line. Please ensure unique key names. If you are using a standard dask collections, consider releasing all the data before resubmitting another computation. More details and help can be found at https://github.com/dask/dask/issues/9888.
Debugging information
---------------------
old task state: processing
old run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDD7A6C0>, key=BasicIndexer((slice(None, None, None),))))),), {})
new run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDFF9540>, key=BasicIndexer((slice(None, None, None),))))),), {})
old token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['8473dc8afac710c0d78f4dadf0eb1b8ef9fe5c46'])]), ('dict', [])])
new token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['05fe405753166f125559e7c9ac558654f107c7e9'])]), ('dict', [])])
old dependencies: set()
new dependencies: set()

WARNING | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-a-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`. This can cause failures and deadlocks down the line. Please ensure unique key names. If you are using a standard dask collections, consider releasing all the data before resubmitting another computation. More details and help can be found at https://github.com/dask/dask/issues/9888.
Debugging information
---------------------
old task state: processing
old run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDD7A6C0>, key=BasicIndexer((slice(None, None, None),))))),), {})
new run_spec: (<function execute_task at 0x00000232EB1D9120>, (ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDFF9540>, key=BasicIndexer((slice(None, None, None),))))),), {})
old token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['8473dc8afac710c0d78f4dadf0eb1b8ef9fe5c46'])]), ('dict', [])])
new token: ('tuple', [('913ceb5b5beb463a9010ec0790bc30002ca34164', []), ('tuple', [('673b402c9db84ac3e891ca3ce3c9a8b9e901612a', ['05fe405753166f125559e7c9ac558654f107c7e9'])]), ('dict', [])])
old dependencies: set()
new dependencies: set()

 DEBUG  | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-y-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`.
 DEBUG  | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-y-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`.
 DEBUG  | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-y-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`.
 DEBUG  | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-y-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`.
 DEBUG  | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-y-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`.
 DEBUG  | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-y-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`.
 DEBUG  | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-y-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`.
 DEBUG  | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-y-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`.
 DEBUG  | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-y-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`.
 DEBUG  | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-y-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`.
 DEBUG  | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-y-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`.
 DEBUG  | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-y-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`.
 DEBUG  | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-y-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`.
 DEBUG  | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-y-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`.
 DEBUG  | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-y-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`.
 DEBUG  | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-y-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`.
 DEBUG  | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-y-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`.
 DEBUG  | distributed.scheduler               | Detected different `run_spec` for key 'original-open_dataset-y-c34960ff895a9a4a2d1c966c401b7d40' between two consecutive calls to `update_graph`.
```

### Anything else we need to know?

I wish I could do one of these:

1. `dataset.load(key=my_unique_key)`
2. `dataset.load(name=my_unique_key)`
3. `dataset.load(pure=False)` to generate a random key

Other workarounds?

### Environment

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.11.5 | packaged by Anaconda, Inc. | (main, Sep 11 2023, 13:26:23) [MSC v.1916 64 bit (AMD64)]
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 85 Stepping 7, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: ('English_United States', '1252')
libhdf5: 1.12.1
libnetcdf: None

xarray: 2024.6.0
pandas: 2.0.3
numpy: 1.24.3
scipy: 1.13.0
netCDF4: None
pydap: None
h5netcdf: None
h5py: 3.9.0
zarr: 2.18.2
cftime: None
nc_time_axis: None
iris: None
bottleneck: 1.3.5
dask: 2024.7.1
distributed: 2024.7.1
matplotlib: 3.8.2
cartopy: None
seaborn: 0.12.2
numbagg: None
fsspec: 2024.6.1
cupy: None
pint: None
sparse: None
flox: None
numpy_groupies: None
setuptools: 68.0.0
pip: 24.1.2
conda: 23.9.0
pytest: 7.4.0
mypy: None
IPython: 8.15.0
sphinx: 5.0.2
</details>

**Comment by templiert** [comment]
If I change [line 332 in `xarray/xarray/backends
/api.py`](https://github.com/pydata/xarray/blob/main/xarray/backends/api.py#L332) from
```python
name_prefix = "open_dataset-"
```
to 
```python
name_prefix = f"open_dataset-{uuid.uuid1()}-"
```
then the `run_spec` warning is not issued any more.

Nevertheless this change did not fix the deadlocks of my larger application from which I extracted this MVCE.

**Comment by dcherian** [comment]
@phofl or @fjetter is this something you can take a look at and suggest what Xarray can/should do. if anything?

**Comment by phofl** [comment]
Ok that's a nasty one. There are multiple small differences

Assume the non-dataclass thing:

```
        futures = [client.submit(read_data, path=path_zarr) for repeat in REPEATS]
        client.gather(futures)
```

You are submitting the same thing twice, Dask identifies this and de-duplicates, we actually run the read_data function only once. Adding a unique key to read_data will make the warning show up as well.

The data class is mutable, so even if you create the data class only once


```
        a = Agent()
        futures = [
            client.submit(a._read_data, path=path_zarr, x=repeat) for repeat in REPEATS
        ]
        client.gather(futures)  # <- run_spec issue with 2 repeats
```

You could mutate the data class between the 2 submits, causing Dask to run things twice. Making the class frozen or removing self from the _read_data method will make the warning go away.

Now why is this warning showing up at all:

So this is an odd one. Every open-air call creates a different ``ZarrArrayWrapper`` class which is embedded into the graph. The wrapper class doesn't have an impact onto the token of the open-dataset key though, so they tokenise to the same thing but the scheduler notices that you have 2 different  ZarrArrayWrapper Classes in there warning you about the conflict.

It looks like the different wrapper classes tokenising to the same thing though, because they ultimately point to the same object. So even if you include those into the tokenise, it wouldn't make a difference. The warnings are fine to ignore here as far as I understand, but we should try to make them go away anyway. The tricky part is that (I assume) you wouldn't want to do the work twice @dcherian if someone calls an open Zarr onto the same thing consecutively, is that correct?

**Comment by fjetter** [comment]
> You could mutate the data class between the 2 submits, causing Dask to run things twice. Making the class frozen or removing self from the _read_data method will make the warning go away.

If the class was mutated, it should yield a different key. To me it rather looks like one of the objects passed are not tokenizing deterministically, i.e. likely one of those classes `ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyIndexedArray(array=<xarray.backends.zarr.ZarrArrayWrapper object at 0x00000232EDD7A6C0>, key=BasicIndexer((slice(None, None, None),))))),)`

**Comment by dcherian** [comment]
> To me it rather looks like one of the objects passed are not tokenizing deterministically, i.e. likely one of those classes 

This I can believe, but it shouldn't crash the system, no? It should run, just a bit inefficiently.

**Comment by phofl** [comment]
> To me it rather looks like one of the objects passed are not tokenizing deterministically, i.e. likely one of those classes

Yep this is the reason for the warning on the scheduler.

> This I can believe, but it shouldn't crash the system, no? It should run, just a bit inefficiently.

It should not impact the worklow in this case, no.

> You could mutate the data class between the 2 submits, causing Dask to run things twice. Making the class frozen or removing self from the _read_data method will make the warning go away.

Sorry I probably wasn't very clear with this. The futures of the client.submit call have different keys for the data class version, and I was wondering why, but the data class isn't the only reason. It looks like our serialise changes how the data class tokenises:

```
        a = Agent()
        tokenize(a._read_data)
        tokenize(a._read_data)

        serialize(a._read_data)
        tokenize(a._read_data)

```

results of the 3 tokenise calls:

```
6b4f5796583b2a41e71a431cf3de37a4
6b4f5796583b2a41e71a431cf3de37a4
e1bba1b1ba7d8659e3f09839e1d14760
```

**Comment by phofl** [comment]
> To me it rather looks like one of the objects passed are not tokenizing deterministically

Yep that's the ``ZarrArrayWrapper`` class (falling back to pickle and it's two different objects)

**Comment by fjetter** [comment]
This looks to be a bug in CPython / pickle or dataclasses that messes with our tokenization since pickling an instance mutates the type

![image](https://github.com/user-attachments/assets/4917c9ca-4aed-4770-9a26-a9f2d9bfbee6)

The fact that this attribute now exists causes our hashes to deviate because we have to tokenize/pickle the object by value

**Comment by fjetter** [comment]
Sorry for this "noise". The TLDR is that I think this is a dask bug and I think what xarray is doing is just fine.

**Comment by fjetter** [comment]
With https://github.com/dask/dask/pull/11320 the reproducer from OP is fixed

**Comment by fjetter** [comment]
While the above PR fixes the warning, from all I can tell this is not one of those cases where this warning would genuinely cause problems. If you are running into a deadlock this is something we should look into separately.

**Comment by templiert** [comment]
Many thanks @dcherian, @phofl and @fjetter.
Glad that it still helped you find something wrong upstream in dask.
> If you are running into a deadlock this is something we should look into separately.

Agreed, thanks.
